lib/api/src/js/externals/function.rs

```diff
@@ -402,11 +415,34 @@ impl Function {
             let js_value = param.as_jsvalue(&store.as_store_ref());
             arr.set(i as u32, js_value);
         }
-        let result = js_sys::Reflect::apply(
-            &self.handle.get(store.as_store_ref().objects()).function,
-            &wasm_bindgen::JsValue::NULL,
-            &arr,
-        )?;
+
+        let result = {
+            let mut r;
+            loop {
+                r = js_sys::Reflect::apply(
+                    &self.handle.get(store.as_store_ref().objects()).function,
+                    &wasm_bindgen::JsValue::NULL,
+                    &arr,
+                );
+                let store_mut = store.as_store_mut();
+                if let Some(callback) = store_mut.inner.on_called.take() {
+                    match callback(store_mut) {
+                        Ok(wasmer_types::OnCalledAction::InvokeAgain) => {
+                            continue;
+                        }
+                        Ok(wasmer_types::OnCalledAction::Finish) => {
+                            break;
+                        }
+                        Ok(wasmer_types::OnCalledAction::Trap(trap)) => {
+                            return Err(RuntimeError::user(trap))
+                        }
+                        Err(trap) => return Err(RuntimeError::user(trap)),
+                    }
+                }
+                break;
+            }
+            r?
+        };
```

Comment:
We can move this into the wasix library itself, I think it'll be cleaner

Reply:
Even if we move this, there still needs to be changes made to this main loop or `asyncify` will not work - it needs to be able to detect if the main loop is replaying a stack or recording a stack

Reply:
Note for the future: We agreed to move this logic out of the API after merging the PR and WASIX merges